### PR TITLE
Add environment variables to identify the binary and crate name

### LINKED ISSUE #8251: Environment variable to identify the binary name
Currently, cargo provides an environment variable `CARGO_PKG_NAME` containing the name of the crate. `clap` uses this in `clap::crate_name` to get the name of the crate to use as the program name.

However, this doesn't work for crates that build more than one binary. I'd like to have an environment variable specifying the name of the binary being built, such as `CARGO_BIN_NAME`, which `clap` could use for a more accurate idea of the name of the program.

For example, given the following `Cargo.toml`:

```toml
[package]
name = "cratename"
version = "0.1.0"
edition = "2018"

[[bin]]
name = "binname"
path = "src/main.rs"

[dependencies]
clap = "2.33.1"
```

When building `src/main.rs`, cargo should set `CARGO_PKG_NAME=cratename` and `CARGO_BIN_NAME=binname`.

### PR DESCRIPTION
**Pull request by reggaemuffin**
Closes #8251

This adds `CARGO_BIN_NAME` and `CARGO_CRATE_NAME` to rustc/rustdoc process env.

`CARGO_BIN_NAME` is added for binary compilation units, `CARGO_CRATE_NAME` is added for binary and library units. 

The `build::crate_env_vars` test was updated to test for this. The test is currently only checking behavior for the binary compile unit.

Documentation was updated to reflect the added environment variables.

### CODE REVIEW

File: src/doc/src/reference/environment-variables.md [hunk 1]
```diff
@@ -184,6 +184,8 @@ let version = env!("CARGO_PKG_VERSION");
 * `CARGO_PKG_DESCRIPTION` — The description from the manifest of your package.
 * `CARGO_PKG_HOMEPAGE` — The home page from the manifest of your package.
 * `CARGO_PKG_REPOSITORY` — The repository from the manifest of your package.
+* `CARGO_CRATE_NAME` — The name of the crate that is currently being compiled.
+* `CARGO_BIN_NAME` — The name of the binary that is currently being compiled (if it is a binary).
```
Comment (ehuss) [inline src/doc/src/reference/environment-variables.md:188]: Maybe this should clarify that on Windows the name does *not* contain the `.exe` extension?

**Comment by ehuss** [comment]
Thanks! This looks good to me.  Just want to get final OK from the team:

@rfcbot fcp merge

**Comment by joshtriplett** [comment]
Looks great, thank you!

**Comment by alexcrichton** [comment]
One possible alternative to `BIN_NAME` might be `BIN_STEM` since "filestem" is typically already known to not include file extensions.

**Comment by reggaemuffin** [comment]
> One possible alternative to `BIN_NAME` might be `BIN_STEM` since "filestem" is typically already known to not include file extensions.

Happy to amend the pr if we have a consensus on the name to go forward with. I personally find filestem a rather unknown word that has a lot in common with filesystem, so while technically being more precise it might be more confusing for the average user, so I would prefer clarifying in the docs and leave it as bin.

**Comment by ehuss** [comment]
I don't have a strong opinion one way or the other.  Unless @alexcrichton wants it to change, I would just leave it.

**Comment by alexcrichton** [comment]
Nah yeah that's a good point, I think just clarifying that the name doesn't include the extension should do the trick!

File: src/doc/src/reference/environment-variables.md (hunk 1 above)
Comment (joshtriplett) [reply src/doc/src/reference/environment-variables.md:188]: Adding "This name does not include any file extension, such as `.exe`." would work.

**Comment by joshtriplett** [comment]
Agreed. Please update that one line in the documentation; I supplied some text that should work.

**Comment by joshtriplett** [comment]
:+1:

**Comment by A1-Triard** [comment]
I've tried last nighly and it looks like `env!("CARGO_CRATE_NAME")` is always set to `"build_script_build"`, `env!("CARGO_BIN_NAME")` does not compile, and `env::var("CARGO_CRATE_NAME")` as well as `env::var("CARGO_BIN_NAME") are not set at all.

**Comment by ehuss** [comment]
@A1-Triard `CARGO_BIN_NAME` is only set while building the binary.

**Comment by A1-Triard** [comment]
@ehuss, I am building a binary. Two binaries in fact. They are not specified in Cargo.toml explicitly, but they lie in right place (src/bin/bin1.rs, src/bin/bin2.rs), thus they automatically detects as binaries.

**Comment by ehuss** [comment]
From your description, it sounds like you are checking from the build script, though?  The following works for me when placed inside the binary:

```rust
// src/main.rs
fn main() {
    println!("Hello, {}", env!("CARGO_BIN_NAME"));
}
```

Are you checking it from the build script, or from the binary?

**Comment by A1-Triard** [comment]
@ehuss, I have understood what you mean. Thank you for explanation.

**Comment by wiknew1** [comment]
This is an essential feature but the documentation for these environment variables are extremely confusing. 

The documentation for `CARGO_BIN_NAME` says "The name of the binary that is **currently being compiled**". From that statement, it's clear that the variable is meant to be checked from the build script. But it's not available there? It's only available FROM inside the binary, AFTER being compiled? 🤷🏽 That doesn't make sense to me at all.

I have multiple bins and only one of them needs the build script. But I cannot check which binary is being compiled from the build script to do that. The only way I found so far is to manually set an env var before running the build and repeating that for each binary I have to build. That's tiresome. Why not have the `CARGO_BIN_NAME` available, where it's actually needed? I've also seen https://github.com/rust-lang/cargo/issues/1430 which was never resolved.

**Pull request merged**